ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -221,6 +205,35 @@ public void stop(final QueryId queryId, final boolean resetOffsets) {
     log.info("Query {} was stopped successfully", queryId);
   }
 
+  private void tryCleanupTopology(QueryId queryId, boolean resetOffsets) {
+    for (int retries = 0; retries < 3; retries++) {
+      try {
+        for (KafkaFuture<Void> toAdd : topolgogiesToAdd) {
+          toAdd.get();
+        }
+        topolgogiesToAdd.clear();
+        kafkaStreams.removeNamedTopology(queryId.toString(), resetOffsets)
+            .all()
+            .get(shutdownTimeout, TimeUnit.MILLISECONDS);
+        if (resetOffsets) {
+          kafkaStreams.cleanUpNamedTopology(queryId.toString());
+        }
+        break;
+      } catch (final TimeoutException | ExecutionException | InterruptedException e) {
+        if (!(e instanceof TimeoutException) || ++retries == 3) {
```

Comment:
I think we increment retries here and in the loop header. Can you add a unit test to make sure the retry method works as intended?

Reply:
Good catch! Test added